Fix: (plz--coding-system) Extract charset from media type

josephmturner opened the pull request:
Previously, plz--coding-system always returned nil since coding-system-from-name expects a coding system name string like "UTF-8".

josephmturner committed:
Fix: (plz--coding-system) Extract charset from media type

Previously, plz--coding-system always returned nil since
coding-system-from-name expects a coding system name string like
"UTF-8".

Co-developed-by: Basil L. Contovounesios <basil@contovou.net>

basil-conto commented on plz.el:
```diff
@@ -961,10 +961,13 @@ Assumes that point is at beginning of HTTP response."
-      (coding-system-from-name content-type))))
+  (when-let* ((headers (or headers (plz--headers)))
+              (raw-content-type (alist-get 'content-type headers))
+              (content-type (downcase raw-content-type))
```
Alternatively, you could bind `case-fold-search` around the `string-match`; saves a string allocation.

basil-conto replied:
Yet more alternatively (for more brownie points?), did you consider using the `mail-parse` library?  See [`(info "(emacs-mime) Interface Functions")`](https://gnu.org/s/emacs/manual/html_node/emacs-mime/Interface-Functions.html).  For example:

```el
(require 'mail-parse)
(mail-content-type-get
 (mail-header-parse-content-type "text/javascript; charset=ISO-8859-1")
 'charset)
;; => "ISO-8859-1"
```

alphapapa replied:
> Alternatively, you could bind `case-fold-search` around the `string-match`; might save a string allocation.

Thanks, that's a good idea.  @josephmturner Would you like to add that, or should I?

alphapapa replied:
> Yet more alternatively (for more brownie points?), did you consider using the built-in `mail-parse` library?

No, I wasn't aware of that function.  Given the simplicity of this scenario, I think I'd prefer to not have to load that library.  Looking at the `rfc2231` library it is aliased to, while that library is certainly useful, the function `rfc2231-parse-string` which gets called looks large, complicated, and would probably be overkill for this.  I can't imagine it would help performance to use that function, either.  Thanks for pointing it out, though.  :)

basil-conto replied:
> I can't imagine it would help performance to use that function

Neither does making HTTP requests ;).

josephmturner replied:
Good idea @basil-conto !  I made this change, rebased, and force-pushed.  Thanks!  I credited you in the commit message.

basil-conto reviewed (approved):
Just pinging in support of this patch because `plz` does not currently decode non-UTF-8 responses, despite claiming to.

To illustrate with a real example, `plz--coding-system` fails to extract the correct charset from a header like `(content-type . "text/javascript; charset=ISO-8859-1")`.

alphapapa commented:
> Just pinging in support of this patch because `plz` does not currently decode non-UTF-8 responses, despite claiming to.
> 
> To illustrate with a real example, `plz--coding-system` fails to extract the correct charset from a header like `(content-type . "text/javascript; charset=ISO-8859-1")`.

@basil-conto  Thank you for reminding me.  Would you be willing to contribute a test case for this?  That would make it easy to apply this as a fix to the stable release branch.

basil-conto commented:
> @basil-conto Thank you for reminding me. Would you be willing to contribute a test case for this?

Sure, but I'd appreciate some pointers: were you thinking of a test that gets `httpbin` to return something non-UTF-8-encoded? (Does it even support that? postmanlabs/httpbin#427)
Or just a unit test for extracting the correct `charset` from a `Content-Type` header?
Or something else?

alphapapa commented:
@basil-conto Good question, of course.  Ideally we'd have both a unit test for the header parsing, and a test that actually receives non-UTF8-encoded content and verifies that it is decoded properly.  If httpbin doesn't make that possible, we might have to get by with a unit test for the header, and a manual test of the functionality, I guess.  :)

josephmturner commented:
Thanks for the review @basil-conto @alphapapa !  I added `case-fold-search` and rebased on top of master.

I think a unit test would be good enough, since the usage of `plz--coding-system` is pretty simple everywhere.

josephmturner commented:
CI for `27` and `snapshot` are failing.   `snapshot` is complaining about `when-let` being obsolete, which is resolved by #71.  The issue with `27` appears more extensive, but since the CI for #71 fails in the same way.  I think this PR is probably fine to merge before addressing that issue.

alphapapa commented:
> Thanks for the review @basil-conto @alphapapa ! I added `case-fold-search` and rebased on top of master.
> 
> I think a unit test would be good enough, since the usage of `plz--coding-system` is pretty simple everywhere.

Thanks, Joseph.



> CI for `27` and `snapshot` are failing. `snapshot` is complaining about `when-let` being obsolete, which is resolved by #71. The issue with `27` appears more extensive, but since the CI for #71 fails in the same way. I think this PR is probably fine to merge before addressing that issue.

Yes, something has changed outside of this repository that's causing the error, `Package ‘emacs-28.1’ is unavailable`. That causes the setup to fail, which causes the Docker image with httpbin to not be started, which causes the tests to fail.  This is happening even though nothing in `plz` or `makem.sh` that I can find has any reference to Emacs 28; `plz` itself depends on no other packages.  

So I don't know what's going on there, and it probably won't be fun to dig into.  The only guess I have is something in https://github.com/purcell/setup-emacs/, which we depend on to install Emacs, but I don't see any obvious candidates there, either.

josephmturner commented:
I'm not sure either.

basil-conto commented:
> Yes, something has changed outside of this repository that's causing the error, `Package ‘emacs-28.1’ is unavailable`.

Emacs emits this error message when one tries to `package-install` dependencies that cannot be satisfied.

> That causes the setup to fail, which causes the Docker image with httpbin to not be started, which causes the tests to fail. This is happening even though nothing in `plz` or `makem.sh` that I can find has any reference to Emacs 28; `plz` itself depends on no other packages.

Looking into the `SANDBOX_DIR`, I see an `init.el` with
```el
(package-selected-packages '(elsa package-lint relint))
```
And indeed these packages seem to be requested by `makem.sh`:
https://github.com/alphapapa/makem.sh/blob/765019a13a2e0ccfe1abdda77fbe1217d2d3bf3f/makem.sh#L637-L640

Not sure if this is the reason, but `elsa` lists [`lsp-mode` as a dependency](https://github.com/emacs-elsa/Elsa/blob/4ce89b0cf313143cd6b7614f4f50aa8179354df6/elsa.el#L9), which in turn [requires Emacs 28](https://github.com/emacs-lsp/lsp-mode/blob/7c0df125c19536646634dca9e27ecb3705db7521/lsp-mode.el#L7).

I'm not familiar with `makem.sh`, so I can't say whether `plz`'s use of `makem.sh` should be configured to avoid installing new linters on old Emacs versions in CI, or whether `makem.sh` should be (made) aware of this to begin with.
